src/ScriptEditor/ui/StatusBar.tsx

```diff
@@ -0,0 +1,226 @@
+  ) {
+    this.node = node;
+    this.editor = editor;
+
+    this.rerender = rerender;
+
+    this.onCloseHandler = null;
+    this.onKeyDownHandler = null;
+    this.onKeyUpHandler = null;
+
+    this.update();
+  }
+
+  // this is used to change the mode shown in the status bar.
+  setMode(ev: { mode: string; subMode?: string }) {
+    if (ev.mode === "visual") {
+      if (ev.subMode === "linewise") {
+        this.mode = "--VISUAL LINE--";
+      } else if (ev.subMode === "blockwise") {
+        this.mode = "--VISUAL BLOCK--";
+      } else {
+        this.mode = "--VISUAL--";
+      }
+    } else {
+      this.mode = `--${ev.mode.toUpperCase()}--`;
+    }
+
+    this.update();
+  }
+
+  // this is used to set the current operator, like d, r, etc or the count, like 5j, 3w, etc.
+  setKeyBuffer(key: string) {
+    this.buffer = key;
+
+    this.update();
+  }
+
+  // this is used to set the input box.
+  setSec(
+    // this is the created HTML element from monaco-vim. We're not going to use it, so it is marked as unused.
+    __text: HTMLElement,
+    // this is used to close the input box and set the cursor back to the line in the monaco-editor. query is the text in the input box.
+    onClose: (query: string) => void,
+    options: {
+      // This handles ESC, Backspace when input is empty, CTRL-C and CTRL-[. query is the text in the input box. close is a function that closes the input box. e is the key event.
+      onKeyDown: (e: KeyboardEvent, query: string, close: () => void) => void;
+      // This handles all other key events. query is the text in the input box. close is a function that closes the input box. e is the key event.
+      onKeyUp: (e: KeyboardEvent, query: string, close: () => void) => void | undefined;
+      // this is a default value for the input box. The box should be empty if this is not set.
+      value: string | undefined;
+    },
+  ) {
+    this.onCloseHandler = onClose;
+    this.onKeyDownHandler = options.onKeyDown;
+    this.onKeyUpHandler = options.onKeyUp ?? null;
+    this.inputValue = options.value || "";
+    this.showInput = true;
+
+    this.update();
+  }
+
+  // this is used to toggle showing the status bar.
+  toggleVisibility(toggle: boolean) {
+    if (toggle) {
+      this.node.current = this.StatusBar();
+    } else {
+      this.node.current = null;
+    }
+  }
+
+  // this is used to close the input box.
+  closeInput = () => {
+    this.showInput = false;
+    this.update();
+    this.editor.focus();
+  };
+
+  // this is used to clean up the status bar on unmount.
+  clear = () => {
+    this.node.current = null;
+  };
+
+  // this is used to show notifications. In game, these won't be rendered, so the function is empty.
+  showNotification(__text: HTMLElement) {
+    return;
+  }
+
+  update = () => {
+    this.node.current = this.StatusBar();
+    this.rerender();
+  };
+
+  keyUp = (e: React.KeyboardEvent) => {
+    const event = new KeyboardEvent("keyup", {
```

Comment:
Can you just do `{...e}`? Do you need to filter out these specific fields?

Reply:
Sadly, I haven't found another way to port the `React.KeyboardEvent` to `KeyboardEvent`. There are over 200 entries that aren't on `KeyboardEvent`, so I filtered out *all* of the necessary properties the package uses.

Reply:
Why are you using KeyboardEvent at all instead of just keeping it as a React.KeyboardEvent?

Reply:
Well, the package uses the simple `KeyboardEvent`. I haven't tested simply passing a `React.KeyboardEvent`. Maybe `React.KeyboardEvent` is based on `KeyboardEvent`? If it works, it works.
Looking at the source code of the package, the only "weird" properties it uses are `e.target.selectionStart` and `e.target.selectionEnd`. Others, like `e.target.value` or `e.key`, are probably on `React.KeyboardEvent`.

Reply:
Similar to Snarling, I don't see/understand why it would be an issue to have extra properties.

Reply:
I'll try out just passing the event and test for any weird behaviors.

Reply:
Good thing we discussed this: The current implementation as well as the one with just passing the event throws error for one specific thing, the `q:` operator, or whatever it is called. When pressing this combination, in the current deployment, a text field opens and no errors are thrown when trying to navigate previous commands. However, nothing happens. No idea why or how it is intended to look, but it just doesn't work.
I'll push the changes necessary to avoid the errors for now, but they won't fix this command not working.